dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProverWithTracker.java

```diff
@@ -0,0 +1,144 @@
+package com.dat3m.dartagnan.encoding;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import org.sosy_lab.java_smt.api.*;
+import org.sosy_lab.java_smt.api.SolverContext.ProverOptions;
+
+import com.google.common.collect.ImmutableMap;
+
+public class ProverWithTracker implements AutoCloseable {
+
+    private final FormulaManager fmgr;
+    private final ProverEnvironment prover;
+    private final String fileName; 
+
+    public ProverWithTracker(SolverContext ctx, String fileName, ProverOptions... options) {
+        this.fmgr = ctx.getFormulaManager();
+        this.prover = ctx.newProverEnvironment(options);
+        this.fileName = fileName;
+        init();
+    }
+
+    // An empty filename means there is no need to dump the encoding
+    private boolean dump() {
+        return !fileName.isEmpty();
+    }
+    
+    private void init() {
+        try {
+            Files.deleteIfExists(Paths.get(fileName));
+            write("(set-logic ALL)\n");
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        removeDuplicatedDeclarations(fileName);
+        prover.close();
+    }
+
+    // Each call to fmgr.dumpFormula(f) defines all variables used in f.
+    // We might end up with several definitions of the declarations.
+    // This might not be the most performant implementation, but duming the 
+    // smtlib2 file is intended only for debugging or trying to understand 
+    // how to optimize the encoding, thus performance is not an issue. 
+    private void removeDuplicatedDeclarations(String fileName) {
```

Comment:
Is the file only valid if it contains no duplicates or is this just to reduce the file?

Reply:
At least both z3 and cvc5 (the solvers I tried) complain.

Reply:
I see. But then I believe you should follow a different approach that only generates the file at the very end (there is limited use for partial/invalid intermediate files).
Also, the way you remove duplicates is error-prone: it will fail if symbols are popped and then redeclared (unlikely to happen in the way we use the solver though).
I think you could just track the sequence of prover calls (push, pop, check, assert(f)) in a list and in the end generate the SMT-lib from that.

Reply:
What is the advantage of writing the file only at the end? Keeping track of everything results is a more "complex" implementation (my first version of the code was doing this) and I don't see the point of this without a clear advantage.

> there is limited use for partial/invalid intermediate files

I disagree about this. The whole reason I implemented this is because I have been waiting for the end of the same verification run for the last two weeks. There are caat iterations that take over 12 hs and I could simply take the intermediate file and check if some other solver is faster (i.e., _is the encoding really hard or just this particular tool struggles about this?_).

Reply:
If you can use the intermediate file it means your smtlib is not invalid, no? That's why I was asking. Or do you run a script to make it valid before testing another solver?

Reply:
Ok, you are right that the scenario I described will not work out of the box with the current implementation. 

If I want to have this, I could either call `removeDuplicatedDeclarations` at the end of `write` (decreases performance) or keep track of all declared variables (increases memory foot print). Since this functionality is mostly for debugging, I don't mind the performance hit and I would be inclined to this solution.

Reply:
I don' think tracking variables is really expensive. You can traverse formulas, so technically you only need a single formula (a single pointer) that contains all variables and then just traverse it. Anyway, a correct implementation does not get around tracking the stack of push/pop operations: you need to do it at least within the duplication deletion phase.
Well, or you just hope that the interaction we produce with the solver never produces a history on which your algorithm breaks (which might be a fair assumption).